File: Cpp/decorator.cpp
#include <iostream>
#include <string>

class Text {
public:
    virtual ~Text() {}
    virtual std::string getText() = 0;  // Pure virtual function
};

class PlainText : public Text {
public:
    PlainText(std::string text) : text(text) {}
    std::string getText() override {
        return text;
    }
private:
    std::string text;
};

class TextDecorator : public Text {
public:
    TextDecorator(Text* text) : text(new PlainText(text->getText())) {}
    ~TextDecorator() { delete text; }
protected:
    Text* text;
};

class BoldText : public TextDecorator {
public:
    BoldText(Text* text) : TextDecorator(text) {}
    std::string getText() override {
        return "<b>" + text->getText() + "</b>";
    }
};

class ItalicText : public TextDecorator {
public:
    ItalicText(Text* text) : TextDecorator(text) {}
    std::string getText() override {
        return "<i>" + text->getText() + "</i>";
    }
};

int main() {
    Text* text = new PlainText("Hello, World!");

    Text* boldText = new BoldText(new PlainText(text->getText()));
    std::cout << boldText->getText() << std::endl;  // Prints: <b>Hello, World!</b>
    delete boldText;

    Text* italicText = new ItalicText(new PlainText(text->getText()));
    std::cout << italicText->getText() << std::endl;  // Prints: <i>Hello, World!</i>
    delete italicText;

    Text* boldItalicText = new BoldText(new ItalicText(new PlainText(text->getText())));
    std::cout << boldItalicText->getText() << std::endl;  // Prints: <b><i>Hello, World!</i></b>
    delete boldItalicText;

    delete text;

    return 0;
}

File: Python/decorator.py
from abc import ABC, abstractmethod


class Text(ABC):
    @abstractmethod
    def get_text(self):
        pass


class PlainText(Text):
    def __init__(self, text):
        self.text = text

    def get_text(self):
        return self.text


class TextDecorator(Text):
    def __init__(self, text: Text):
        self.text = text


class BoldText(TextDecorator):
    def get_text(self):
        return f'<b>{self.text.get_text()}</b>'


class ItalicText(TextDecorator):
    def get_text(self):
        return f'<i>{self.text.get_text()}</i>'


if __name__ == "__main__":

    # Użycie
    text = PlainText('Hello, World!')

    bold_text = BoldText(text)
    print(bold_text.get_text())  # Wyświetla: <b>Hello, World!</b>

    italic_text = ItalicText(text)
    print(italic_text.get_text())  # Wyświetla: <i>Hello, World!</i>

    bold_italic_text = BoldText(ItalicText(text))
    print(bold_italic_text.get_text())  # Wyświetla: <b><i>Hello, World!</i></b>
